lib/Dialect/ArithExt/IR/ArithExtDialect.cpp

```diff
@@ -0,0 +1,46 @@
+#include "lib/Dialect/ArithExt/IR/ArithExtDialect.h"
+
+#include "mlir/include/mlir/Dialect/Arith/IR/Arith.h"    // from @llvm-project
+#include "mlir/include/mlir/IR/DialectImplementation.h"  // from @llvm-project
+
+// NOLINTNEXTLINE(misc-include-cleaner): Required to define ArithExtOps
+#include "lib/Dialect/ArithExt/IR/ArithExtOps.h"
+
+// Generated definitions
+#include "lib/Dialect/ArithExt/IR/ArithExtDialect.cpp.inc"
+
+#define GET_OP_CLASSES
+#include "lib/Dialect/ArithExt/IR/ArithExtOps.cpp.inc"
+
+namespace mlir {
+namespace heir {
+namespace arith_ext {
+
+void ArithExtDialect::initialize() {
+  addOperations<
+#define GET_OP_LIST
+#include "lib/Dialect/ArithExt/IR/ArithExtOps.cpp.inc"
+      >();
+}
+
+LogicalResult BarrettReduceOp::verify() {
+  auto inputType = getInput().getType();
+  unsigned bitWidth;
+  if (auto tensorType = dyn_cast<RankedTensorType>(inputType)) {
+    bitWidth = tensorType.getElementTypeBitWidth();
+  } else if (auto integerType = dyn_cast<IntegerType>(inputType)) {
+    bitWidth = integerType.getWidth();
+  }
+  auto cmod = APInt(64, getModulus());
+  if (bitWidth != 2 * (cmod - 1).getActiveBits()) {
+    return emitOpError()
+           << "input bitwidth is required to be 2w, where w "
+              "is the smallest bit-width that contains the range [0, modulus).";
+  }
```

Comment:
Should we bound this instead of computing an exact inequality? `bitWidth <= 2 * w`?

Might also be worth outputting the expected 2w bitwidth exactly in the error (e.g. expected 10 bits, got 11)

Reply:
Yep, I think we could allow for the range `w <= bitWidth <= 2 * w`, as if we have a bit-width less than `w` we may truncate the result.